### PR TITLE
Enable running TOSA reference_model e2e tests on executorch public CI

### PR DESCRIPTION
**Pull request by Jerry-Ge**
Enable running TOSA ref_model e2e tests and some refactoring

### CODE REVIEW

**Comment by Jerry-Ge** [comment]
ciflow/trunk

File: examples/arm/run.sh
Comment (digantdesai) [inline examples/arm/run.sh:176]: These are effectively different lowering and execution paths, and since they are in the same script with `-e` if something fails before this, then this won't run. Do you want to either selectively apply `-e` or split this into running different individual scripts so we can get more complete signal when something fails?

Comment (Jerry-Ge) [reply examples/arm/run.sh:176]: I agree we can split into separate files (`run.sh` and `run_tosa_e2e.sh`) which will be much cleaner. If this sounds good, I will push the changes. 

Another question is: do i need to do anything else to trigger the running of the new shell script `run_tosa_e2e.sh` ?

Comment (robell) [reply examples/arm/run.sh:176]: hopefully we can add another backend under https://github.com/pytorch/executorch/blob/c44dadf2da17a160768b5aac2e5450d2c835616f/.github/workflows/trunk.yml#L140
Or if preferred from the CI flow we can add it as a script.

I think we should restructure this a bit; I view things as:
  `examples/arm/` are typically user run scripts and tutorials
  `backends/arm/test/*` are unit tests and test scripts

We should add `backends/arm/test/test.sh which tests both the example (as we don't want it to regress either) and the TOSA AoT flow.

Comment (digantdesai) [reply examples/arm/run.sh]: Agree, we should have these scripts under backends/arm.

Comment (robell) [reply examples/arm/run.sh]: Thanks Jerry, changes look sensible to me. being nit-picky having the same name for the invocation script and the .py would be cleaner? might be easier if we pattern match it in automation for some reason later too.

Don't think it's worth doing now, but we might want to rename slightly when we have more test flows, as we have test_tosa.py (unit tests which test edge to tosa) and arm_tosa_reference.py which captures the intermediate compiler output (tosa fb, usually a wire format in AoT) and runs it on the reference.  so being consistent as we add more and looking to something like:

for the AoT compiler flows (by the backend options (which runtime target) that are passed):
unit_tosa
unit_vela
unit_<other/many>

test_tosa_reference
test_u55 (e.g. like the examples/arm/ flow but for when we have substantial correctness testing of a u55/65 backend).

Now it's down to how the CI team would prefer to integrate?

Comment (digantdesai) [reply examples/arm/run.sh]: Yeah let me ping Huy. 

Also I am ok with the change overall besides names bikeshedding which we should converge as we develop more.

File: .github/workflows/trunk.yml
Comment (huydhn) [inline .github/workflows/trunk.yml:177]: This looks like a typo, I think you mean

```suggestion
        PYTHON_EXECUTABLE=python bash backends/arm/test/arm_tosa_reference.sh
```

File: .github/workflows/trunk.yml
Comment (huydhn) [reply .github/workflows/trunk.yml]: As this just needs a common `linux.2xlarge` runner, let's keep them separate for now.  On the other hand, having too many tiny jobs that takes less than 15 to finish is a waste, so we might want to consolidate them into bigger a bigger chunk.  But that's an optimization step that could be done independently of this change IMO.

File: .github/workflows/trunk.yml
Comment (Jerry-Ge) [reply .github/workflows/trunk.yml:177]: ah, yes.

**Review by huydhn — APPROVED** [review]

**Comment by huydhn** [comment]
The CI change looks fine, but the job is still failing due to a minor issue with the wrong path here https://github.com/pytorch/executorch/actions/runs/7212751191/job/19651117657?pr=1335

**Comment by Jerry-Ge** [comment]
> The CI change looks fine, but the job is still failing due to a minor issue with the wrong path here https://github.com/pytorch/executorch/actions/runs/7212751191/job/19651117657?pr=1335

yep, let me fix this quickly.

**Review by digantdesai — APPROVED** [review]
Let me know when ready and I can pull this in. @Jerry-Ge

**Comment by Jerry-Ge** [comment]
> Let me know when ready and I can pull this in. @Jerry-Ge

@digantdesai and @huydhn , seems we're still missing one dependency `flatc` here: https://github.com/pytorch/executorch/actions/runs/7213201076/job/19652543330?pr=1335#step:11:471

**Comment by huydhn** [comment]
I see that flatc is installed in /opt/conda/envs/py_3.10/bin/flatc https://github.com/pytorch/executorch/actions/runs/7213201076/job/19652543330?pr=1335#step:11:259, so may be the path of the input file is wrong `unable to load file: ./backends/arm/third-party/serialization_lib/schema/tosa.fbs`?

**Comment by Jerry-Ge** [comment]
> I see that flatc is installed in /opt/conda/envs/py_3.10/bin/flatc https://github.com/pytorch/executorch/actions/runs/7213201076/job/19652543330?pr=1335#step:11:259, so may be the path of the input file is wrong `unable to load file: ./backends/arm/third-party/serialization_lib/schema/tosa.fbs`?

yes, that's right. updated the path and let's wait for the new run to finish.

**Comment by Jerry-Ge** [comment]
> > I see that flatc is installed in /opt/conda/envs/py_3.10/bin/flatc https://github.com/pytorch/executorch/actions/runs/7213201076/job/19652543330?pr=1335#step:11:259, so may be the path of the input file is wrong `unable to load file: ./backends/arm/third-party/serialization_lib/schema/tosa.fbs`?
> 
> yes, that's right. updated the path and let's wait for the new run to finish.

@huydhn hmm, i think the updated relative path should be correct? https://github.com/pytorch/executorch/actions/runs/7213501162/job/19653512726?pr=1335#step:11:461

`arm_tosa_reference.py` is at `backends/arm/test/arm_tosa_reference.py`

the schema file is at `/backends/arm/third-party/serialization_lib/schema/tosa.fbs` 

we're using the following path in the python script: `../third-party/serialization_lib/schema/tosa.fbs`

**Comment by Jerry-Ge** [comment]
FYI: still playing/testing the upstream CI, don't merge this right now. @huydhn @digantdesai

**Comment by Jerry-Ge** [comment]
@huydhn seems `tosa_reference_model` is still missing in the PATH? https://github.com/pytorch/executorch/actions/runs/7213874941/job/19654713142?pr=1335#step:11:444

File: .github/workflows/trunk.yml
Comment (huydhn) [inline .github/workflows/trunk.yml:176]: I read the code a bit and I think this is needed to load the correct PATH to `tosa_reference_model` that you need for the job:

```suggestion

  source /opt/arm-sdk/setup_path.sh
```

**Comment by huydhn** [comment]
@Jerry-Ge It's running now, but I'm seeing a core dump crash https://github.com/pytorch/executorch/actions/runs/7228865467/job/19698949725#step:11:440, so something is still not right?

**Comment by Jerry-Ge** [comment]
> @Jerry-Ge It's running now, but I'm seeing a core dump crash https://github.com/pytorch/executorch/actions/runs/7228865467/job/19698949725#step:11:440, so something is still not right?

nice and thanks! I can see the issue is that: 

- We're generating the tosa flatbuffer using 0.90.0d
- While the serialization_lib used in reference_model/thirdparty is still at 0.80.0 https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/+/refs/heads/v0.80/thirdparty/

We probably need to bump the reference_model version to HEAD to fix the error: https://review.mlplatform.org/admin/repos/tosa/reference_model,branches

@huydhn

**Comment by Jerry-Ge** [comment]
> > @Jerry-Ge It's running now, but I'm seeing a core dump crash https://github.com/pytorch/executorch/actions/runs/7228865467/job/19698949725#step:11:440, so something is still not right?
> 
> nice and thanks! I can see the issue is that:
> 
>     * We're generating the tosa flatbuffer using 0.90.0d for serialization_lib
> 
>     * While the serialization_lib used in reference_model/thirdparty is still at 0.80.0 https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/+/refs/heads/v0.80/thirdparty/
> 
> 
> We probably need to bump the reference_model version to HEAD to fix the error: https://review.mlplatform.org/admin/repos/tosa/reference_model,branches
> 
> https://github.com/pytorch/executorch/blob/f836d2d193c88b6f33f054ce3087da90cd59a434/examples/arm/setup.sh#L170C26-L170C26
> 
> @huydhn

I updated the reference_model branch to `main` in `setup.sh`, seems it's still running on the v80.0 branch? @huydhn

**Comment by Jerry-Ge** [comment]
> > > @Jerry-Ge It's running now, but I'm seeing a core dump crash https://github.com/pytorch/executorch/actions/runs/7228865467/job/19698949725#step:11:440, so something is still not right?
> > 
> > 
> > nice and thanks! I can see the issue is that:
> > ```
> > * We're generating the tosa flatbuffer using 0.90.0d for serialization_lib
> > 
> > * While the serialization_lib used in reference_model/thirdparty is still at 0.80.0 https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/+/refs/heads/v0.80/thirdparty/
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We probably need to bump the reference_model version to HEAD to fix the error: https://review.mlplatform.org/admin/repos/tosa/reference_model,branches
> > https://github.com/pytorch/executorch/blob/f836d2d193c88b6f33f054ce3087da90cd59a434/examples/arm/setup.sh#L170C26-L170C26
> > @huydhn
> 
> I updated the reference_model branch to `main` in `setup.sh`, seems it's still running on the v80.0 branch? @huydhn

@huydhn happy new year! could you help take another look on this PR : )

**Comment by guangy10** [comment]
> > > @Jerry-Ge It's running now, but I'm seeing a core dump crash https://github.com/pytorch/executorch/actions/runs/7228865467/job/19698949725#step:11:440, so something is still not right?
> > 
> > 
> > nice and thanks! I can see the issue is that:
> > ```
> > * We're generating the tosa flatbuffer using 0.90.0d for serialization_lib
> > 
> > * While the serialization_lib used in reference_model/thirdparty is still at 0.80.0 https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/+/refs/heads/v0.80/thirdparty/
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We probably need to bump the reference_model version to HEAD to fix the error: https://review.mlplatform.org/admin/repos/tosa/reference_model,branches
> > https://github.com/pytorch/executorch/blob/f836d2d193c88b6f33f054ce3087da90cd59a434/examples/arm/setup.sh#L170C26-L170C26
> > @huydhn
> 
> I updated the reference_model branch to `main` in `setup.sh`, seems it's still running on the v80.0 branch? @huydhn

@Jerry-Ge The `setup.sh` is copied via the build script ([here](https://github.com/pytorch/executorch/blob/main/.ci/docker/build.sh#L43-L44)) and used by the docker image ([here](https://github.com/pytorch/executorch/blob/main/.ci/docker/ubuntu/Dockerfile#L69-L72)), so I think you only need to update the scrip to `0.90.0d`.

**Comment by huydhn** [comment]
> @huydhn happy new year! could you help take another look on this PR : )

Happy New Year!  And sorry for missing your message last week.  I'm updating the docker-build workflow a bit to correctly rebuild the Docker image used by the job.  Let's see if this works now.

Also, the base commit is rather old, so please help do a rebase to latest ExecuTorch main commit.  I realize that my `@pytorchbot rebase -b main` command doesn't work on ExecuTorch PR yet and will create an issue to fix that.

File: examples/arm/setup.sh
Comment (huydhn) [inline examples/arm/setup.sh:170]: I see only [v0.90a0](https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/+/refs/tags/v0.90a0) on https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/, so the build is failing with `Remote branch 0.90.0d not found in upstream origin`.  Let me try to update the tag here to the correct value and rerun CI.

```suggestion
        git clone https://review.mlplatform.org/tosa/reference_model -b v0.90a0
```

Comment (huydhn) [reply examples/arm/setup.sh]: I tried it out with this `v0.90a0` branch and ended up with a slightly different version mismatch error git clone https://review.mlplatform.org/tosa/reference_model -b 0.90.0d

```
Read flatbuffer version 0.90.0d is not compatible with serializer version 0.80.0d (instead of 0.80.0)
```

cc @guangy10 Do you know where to find that version as it's not listed on https://review.mlplatform.org/plugins/gitiles/tosa/reference_model?

Comment (Jerry-Ge) [reply examples/arm/setup.sh]: Yea, `reference_model:v0.90a0` is still using the `0.80.0d` thirdparty serialization_lib which is causing the issue above. 

I changed to `git clone https://review.mlplatform.org/tosa/reference_model -b main` in the latest commit which is now currently using the 0.90.0d serialization_lib. https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/+/refs/heads/main/thirdparty/serialization_lib

But it's still getting `ERROR: Read flatbuffer version 0.90.0d is not compatible with serializer version 0.80.0d`. Feels like the CI is still picking the v0.80 branch? 

Thanks for the help. This is a little bit convoluted now...

Comment (huydhn) [reply examples/arm/setup.sh]: I suspect this is due to the TODO that I mention in `.ci/docker/build.sh`, so I'm pushing a comment change there to trigger the rebuild.  Once the build is done https://github.com/pytorch/executorch/actions/runs/7453481163, could you help do another rebase or any minor changes to update the PR one more time.  That will pull the latest images from https://github.com/pytorch/executorch/actions/runs/7453481163.

Comment (Jerry-Ge) [reply examples/arm/setup.sh]: yes, for sure. I will try that once the new build is done. Thanks!

**Comment by Jerry-Ge** [comment]
@huydhn this finally runs! thanks. Let me squash the commits then we can merge.

File: examples/arm/setup.sh
Comment (huydhn) [reply examples/arm/setup.sh]: It finally looks correct now https://github.com/pytorch/executorch/actions/runs/7454335305/job/20281469352?pr=1335, but I need to hardcode the Docker image to the correct version from https://github.com/pytorch/executorch/actions/runs/7453481163.  To unblock this change, let me import this and follow up on that Docker issue in a separate PR.  Does this sound good to you?

Comment (huydhn) [reply examples/arm/setup.sh]: Oh, I know a cleaner word around for now.  I could delete the upstream Docker image to ensure that the correct one is found and used instead of hard coding the version.  No action needed on your end though.  Please squash the commits as usual and I will reimport it.  Thank you for your patience :)

Comment (Jerry-Ge) [reply examples/arm/setup.sh]: no problem. thanks

**Comment by Jerry-Ge** [comment]
> @huydhn this finally runs! thanks. Let me squash the commits then we can merge.

squash and rebase done. let me know if i need to do anything else?

**Comment by huydhn** [comment]
> squash and rebase done. let me know if i need to do anything else? when the hard-coded docker image commit was reverted, the error now goes back.

I get it sorted out by deleting the old image https://github.com/pytorch/executorch/actions/runs/7454572246/job/20284947302?pr=1335.  This looks good now.

@digantdesai @guangy10 If this looks ok to you, please help stamp D52173335 to land this, I import the PR, so my approval doesn't count :)

**Comment by digantdesai** [comment]
LGTM, looking at the CI, @Jerry-Ge #1336 broke `test-arm-backend-delegate` job. Can you please fix it? I can create an issue for this if you want.

**Comment by digantdesai** [comment]
@huydhn sorry missed this party, are we planning to use a specific version as oppose to main branch once the Docker issue is fixed?

**Comment by Jerry-Ge** [comment]
> LGTM, looking at the CI, @Jerry-Ge #1336 broke `test-arm-backend-delegate` job. Can you please fix it? I can create an issue for this if you want.

Thanks Digant! Looks like it's a Vela problem for Conv2d and I need to find the right person to fix that. Yea, please create a issue and I will relay that to the right person.

**Comment by Jerry-Ge** [comment]
> @huydhn sorry missed this party, are we planning to use a specific version as oppose to main branch once the Docker issue is fixed?

TOSA 0.90 is about to release soon. Once that's been done, we can change everything to 0.90 then all things should be smooth.

**Comment by huydhn** [comment]
> LGTM, looking at the CI, @Jerry-Ge #1336 broke `test-arm-backend-delegate` job. Can you please fix it? I can create an issue for this if you want.

That failure also happens in trunk atm https://hud.pytorch.org/hud/pytorch/executorch/main/1?per_page=50&name_filter=arm from what I see, so it might be unrelated to the change here and can be fixed separately.  If that's the case, I think we can:

1. Merge this
2. Fix the Docker issue
3. Pin to 0.90 branch once it's ready

File: .github/workflows/docker-builds.yml
Comment (guangy10) [inline .github/workflows/docker-builds.yml:10]: @huydhn IIUC, dock image rebuild will be triggered if anything changed in this dir. Should we narrow the triggering condition down to detecting changes in `examples/arm/setup.sh` and `examples/arm/ethos-u-setup/**` eg reference_model version bump? @Jerry-Ge only these two places matters regarding setup the test env right?

File: .ci/docker/build.sh
Comment (guangy10) [inline .ci/docker/build.sh:46]: @huydhn IIUC, it will re-run this step (copy new setup.sh and its deps) when the docker image rebuild is triggered, in this case when Arm bumps up the reference model version in setup.sh. So if we are tracking the changes in setup.sh, we should get the automatic docker image rebuild as needed, right? Just curious if you indicate something different

File: .github/workflows/docker-builds.yml
Comment (huydhn) [reply .github/workflows/docker-builds.yml:10]: You're right that Docker rebuild will be triggered for all changes in `examples/arm` folder. I could narrow it down to only these places if only they are needed.  Although, may be it's easier to just do a blanket include here.

File: .ci/docker/build.sh
Comment (huydhn) [reply .ci/docker/build.sh:46]: Yeah, the rebuild is done alright, but there is a complication where the Docker image hash tag is computed from the commit hash of the `.ci/docker` directory, the standard location of everything docker-related.  So, what we have here is that the Docker image is rebuilt, but its hash tag remains the same because `examples/arm/` is not under `.ci/docker`, so the job doesn't reupload the image thinking that nothing has changed.  That's the reason why @Jerry-Ge tried to update the branch, but didn't see the new change took effect.

I could apply the same workaround that PyTorch is using here by removing `cp` and creating a softlink.

File: examples/arm/setup.sh
Comment (guangy10) [inline examples/arm/setup.sh:170]: I understand that the change here is a temporary workaround to unblock this PR just because the new version [`0.90.0d` is not available](https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/). I think the right way for future update will still be the version based (stable) instead attempting to detect and build on new hash from main. So the process would be as simple as:
  1. Make a new version available in https://review.mlplatform.org/plugins/gitiles/tosa/reference_model/
  2. Create a PR to update examples/arms/setup.sh to pick up the new version. In that PR, we should expect to see the docker image is rebuilt using the updated version and `test-arm-reference-delegation` is using the updated version
  
cc: @huydhn @Jerry-Ge

File: .github/workflows/docker-builds.yml
Comment (huydhn) [reply .github/workflows/docker-builds.yml:10]: Hmm, thinking a bit more, I think we should keep the rule as it's because it reflects the way the whole directory is copied to Docker `cp -r ../../examples/arm/ ./arm`.

**Review by guangy10 — APPROVED** [review]
@huydhn Thanks for digging into the issue and suggested fixes. This PR looks good to me. Just want to double click on the expected process for future version bumps. I commented it inline and let me know if it makes sense.

File: examples/arm/setup.sh
Comment (huydhn) [reply examples/arm/setup.sh:170]: Agree, pinning the stable version is definitely the way to go on CI once the version is ready.

File: .github/workflows/docker-builds.yml
Comment (guangy10) [reply .github/workflows/docker-builds.yml:10]: @huydhn Remind me if we do need to copy the entire `examples/arm` folder. IIUC, those [*`.py` files and the `run.sh` script](https://github.com/pytorch/executorch/tree/main/examples/arm) are not needed by the docker image. Those files may be updated more frequently than the actual dependency for docker image build. Maybe a better way is to collect all deps that docker image needs in a single dir eg `examples/arm/setup/**` and monitor the changes there for docker image rebuild?

File: .ci/docker/build.sh
Comment (guangy10) [reply .ci/docker/build.sh:46]: Oh I see, the hash here is for the docker image not the hash of the `reference_model`. @huydhn Maybe you could elaborate more about when the docker image hash tag is checked and compared? I understand that once those files are copied, the hash of the docker image won't change, but shouldn't the docker image build be triggered based on .github/workflows/docker-builds.yml where we set to monitor the `examples/arm/` dir. In that case, why is it still needed to check and compare the docker image hash?

File: .github/workflows/docker-builds.yml
Comment (huydhn) [reply .github/workflows/docker-builds.yml:10]: Yeah, putting them altogether in a subdirectory like `examples/arm/setup/**` makes sense to me.  As I'm reorganize these scripts a bit to fix the Docker build part, let me include that part there too

**Comment by guangy10** [comment]
@Jerry-Ge The follow-up items are added:
  - #1568
  - #1569

File: .ci/docker/build.sh
Comment (huydhn) [reply .ci/docker/build.sh:46]: The logic behind that is in this GitHub action https://github.com/pytorch/test-infra/blob/main/.github/actions/calculate-docker-image/action.yml#L161-L163, you can see that the Docker image is not pushed again if the same hash tag already exists.  And the hash tag comes from the content of `.ci/docker` folder itself https://github.com/pytorch/test-infra/blob/main/.github/actions/calculate-docker-image/action.yml#L70

Comment (huydhn) [reply .ci/docker/build.sh:46]: I think this option should help https://github.com/pytorch/test-infra/pull/4866, this force pushes the rebuilt Docker image even if the hash tag already exists.

**Pull request closed without merging**
